File: testdata/fixtures/javascript/safe/command_safe.ts
import { Request, Response } from 'express';
import { execFile } from 'child_process';
import { spawn } from 'child_process';

// SAFE: Using execFile with separate arguments (no shell injection)
export function pingHost(req: Request, res: Response): void {
  const host = req.query.host as string;

  // Validate hostname format (alphanumeric and dots only)
  if (!/^[a-zA-Z0-9.-]+$/.test(host)) {
    res.status(400).json({ error: 'Invalid hostname format' });
    return;
  }

  // SAFE: execFile passes arguments as an array, not through a shell
  execFile('ping', ['-c', '3', host], (error, stdout, stderr) => {
    if (error) {
      res.status(500).json({ error: 'Ping failed' });
      return;
    }
    res.json({ output: stdout });
  });
}

// SAFE: Using spawn with explicit arguments array
export function gitClone(req: Request, res: Response): void {
  const repoUrl = req.body.url;

  // Validate URL format
  if (!/^https:\/\/github\.com\/[\w.-]+\/[\w.-]+\.git$/.test(repoUrl)) {
    res.status(400).json({ error: 'Invalid repository URL' });
    return;
  }

  // SAFE: spawn with array of arguments, no shell interpretation
  const child = spawn('git', ['clone', '--depth', '1', repoUrl, '/tmp/repos/latest'], {
    shell: false,
  });

  let output = '';
  child.stdout.on('data', (data) => { output += data; });
  child.on('close', (code) => {
    res.json({ status: code === 0 ? 'cloned' : 'failed', output });
  });
}

File: testdata/fixtures/go/safe/command_safe.go
package safe

import (
	"net/http"
	"os/exec"
	"regexp"
)

// SAFE: Command execution without shell interpreter - arguments passed separately.
// Should NOT trigger GTSS-INJ-002.

var validHostname = regexp.MustCompile(`^[a-zA-Z0-9][a-zA-Z0-9.-]{0,253}[a-zA-Z0-9]$`)

func HandlePing(w http.ResponseWriter, r *http.Request) {
	host := r.URL.Query().Get("host")

	// Validate hostname format before use
	if !validHostname.MatchString(host) {
		http.Error(w, "invalid hostname", http.StatusBadRequest)
		return
	}

	// Safe: arguments are separate, no shell interpreter involved
	cmd := exec.Command("ping", "-c", "3", host)
	output, err := cmd.CombinedOutput()
	if err != nil {
		http.Error(w, "ping failed", http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "text/plain")
	w.Write(output)
}

func HandleListFiles(w http.ResponseWriter, r *http.Request) {
	// Safe: hardcoded arguments, no user input in command
	cmd := exec.Command("ls", "-la", "/var/log")
	output, _ := cmd.CombinedOutput()
	w.Write(output)
}
